features/mbedtls/targets/TARGET_Silicon_Labs/crypto_aes.c

```diff
@@ -0,0 +1,545 @@
+                    output[processed] = (unsigned char)( c ^ iv[n] );
+                    iv[n] = (unsigned char) c;
+                }
+                n = ( n + 1 ) & 0x0F;
+                processed++;
+            }
+
+            crypto_management_release(device);
+        }
+    }
+
+    if ( iv_off ) {
+        *iv_off = n;
+    }
+
+    return ret;
+}
+
+/*
+ * AES-CFB8 buffer encryption/decryption
+ */
+int mbedtls_aes_crypt_cfb8( mbedtls_aes_context *ctx,
+                            int mode,
+                            size_t length,
+                            unsigned char iv[16],
+                            const unsigned char *input,
+                            unsigned char *output )
+{
+    unsigned char c;
+    unsigned char ov[17];
+    int ret = 0;
+
+    if( ctx == NULL || input == NULL || output == NULL || iv == NULL ) {
+        return ( MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH );
+    }
+
+    if ( ctx->keybits != 128UL && ctx->keybits != 256UL) {
+        return MBEDTLS_ERR_AES_INVALID_KEY_LENGTH;
+    }
+
+    while( length-- )
+    {
+        memcpy( ov, iv, 16 );
+        if ( (ret = mbedtls_aes_crypt_ecb( ctx, MBEDTLS_AES_ENCRYPT, iv, iv ) ) != 0 ) {
+            return ret;
+        }
+
+        if( mode == MBEDTLS_AES_DECRYPT )
+            ov[16] = *input;
+
+        c = *output++ = (unsigned char)( iv[0] ^ *input++ );
+
+        if( mode == MBEDTLS_AES_ENCRYPT )
+            ov[16] = c;
+
+        memcpy( iv, ov + 1, 16 );
+    }
+
+    return ret;
+}
+#endif /*MBEDTLS_CIPHER_MODE_CFB */
+
+#if defined(MBEDTLS_CIPHER_MODE_CTR)
+/*
+ * AES-CTR buffer encryption/decryption
+ */
+int mbedtls_aes_crypt_ctr( mbedtls_aes_context *ctx,
+                           size_t length,
+                           size_t *nc_off,
+                           unsigned char nonce_counter[16],
+                           unsigned char stream_block[16],
+                           const unsigned char *input,
+                           unsigned char *output )
+{
+    size_t n = nc_off ? *nc_off : 0;
+    size_t processed = 0;
+    int ret = 0;
+    CORE_DECLARE_IRQ_STATE;
+
+    if( ctx == NULL || input == NULL || output == NULL || nonce_counter == NULL || stream_block == NULL ) {
+        return ( MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH );
+    }
+
+    if ( ctx->keybits != 128UL && ctx->keybits != 256UL) {
+        return MBEDTLS_ERR_AES_INVALID_KEY_LENGTH;
+    }
+
+    while ( processed < length ) {
+        if ( n > 0 ) {
+            /* start by filling up the IV */
+            output[processed] = (unsigned char)( input[processed] ^ stream_block[n] );
+            n = ( n + 1 ) & 0x0F;
+            processed++;
+            continue;
+        } else {
+            /* process one ore more blocks of data */
+            CRYPTO_TypeDef *device = crypto_management_acquire();
+            device->WAC = 0;
+            device->CTRL = CRYPTO_CTRL_INCWIDTH_INCWIDTH4;
+
+            CORE_ENTER_CRITICAL();
```

Comment:
The resource we are protecting with critical sections is writing to or reading from 128-bit, 256-bit or 512-bit registers in our accelerator. Since that operation is done with actionable reads/writes from/to a single 32-bit register, they have to be protected against preemption.

The other registers (such as WAC and CTRL) are regular registers, which we get control over by acquiring the device. In the case of preemption, these values will be backed up and put back by the preempting code, so their use is safe.